Lib/inspect.py

```diff
@@ -2384,52 +2384,34 @@ def p(name_node, default_node, default=empty):
     return cls(parameters, return_annotation=cls.empty)
 
 
-def _signature_from_builtin(cls, func, skip_bound_arg=True):
+def _signature_from_code(cls,
+                                func_code,
+                                globals=None,
+                                locals=None,
+                                eval_str=False,
+                                is_duck_function=False,
+                                func=None):
     """Private helper function to get signature for
-    builtin callables.
+    code objects.
     """
-
-    if not _signature_is_builtin(func):
-        raise TypeError("{!r} is not a Python builtin "
-                        "function".format(func))
-
-    s = getattr(func, "__text_signature__", None)
-    if not s:
-        raise ValueError("no signature found for builtin {!r}".format(func))
-
-    return _signature_fromstr(cls, func, s, skip_bound_arg)
-
-
-def _signature_from_function(cls, func, skip_bound_arg=True,
-                             globals=None, locals=None, eval_str=False):
-    """Private helper: constructs Signature for the given python function."""
-
-    is_duck_function = False
-    if not isfunction(func):
-        if _signature_is_functionlike(func):
-            is_duck_function = True
-        else:
-            # If it's not a pure Python function, and not a duck type
-            # of pure function:
-            raise TypeError('{!r} is not a Python function'.format(func))
-
-    s = getattr(func, "__text_signature__", None)
-    if s:
-        return _signature_fromstr(cls, func, s, skip_bound_arg)
-
     Parameter = cls._parameter_cls
 
     # Parameter information.
-    func_code = func.__code__
     pos_count = func_code.co_argcount
     arg_names = func_code.co_varnames
     posonly_count = func_code.co_posonlyargcount
     positional = arg_names[:pos_count]
     keyword_only_count = func_code.co_kwonlyargcount
     keyword_only = arg_names[pos_count:pos_count + keyword_only_count]
-    annotations = get_annotations(func, globals=globals, locals=locals, eval_str=eval_str)
-    defaults = func.__defaults__
-    kwdefaults = func.__kwdefaults__
+    if func is not None:
+        annotations = get_annotations(func, globals=globals, locals=locals, eval_str=eval_str)
+        defaults = func.__defaults__
+        kwdefaults = func.__kwdefaults__
+    else:
+        # `func` can be `None` when we get a signature from just a `CodeObject`
```

Comment:
I don't think that this comment is relevant in the _signature_from_code() function.

If you care, add a comment in the "is not None" code path to explain that it's the _signature_from_function() code path?